src/unix/linux.c

```diff
@@ -2281,6 +2281,124 @@ uint64_t uv_get_available_memory(void) {
 }
 
 
+#define BUF_SIZE 1024
+
+static uv__cpu_constraint uv__get_cgroupv2_constrained_cpu(const char* cgroup) {
+  uv__cpu_constraint constraint = {0, 0, 0};
+  char path[256];
+  char buf[BUF_SIZE];
+  unsigned int weight;
+  int cgroup_size;
+  const char* cgroup_trimmed;
+
+  /* Trim ending \n by replacing it with a 0 */
+  cgroup_trimmed = cgroup + sizeof("0::/") - 1;      /* Skip the prefix "0::/" */
+  cgroup_size = (int)strcspn(cgroup_trimmed, "\n");  /* Find the first slash */
+
+  /* Construct the path to the cpu.max file */
+  snprintf(path, sizeof(path), "/sys/fs/cgroup/%.*s/cpu.max", cgroup_size,
+           cgroup_trimmed);
+
+  /* Read cpu.max */
+  if (uv__slurp(path, buf, sizeof(buf)) >= 0) {
+    char max_usage_str[16];
+    sscanf(buf, "%15s %lld", max_usage_str, &constraint.period_length);
+    if (strcmp(max_usage_str, "max") == 0)
+      constraint.quota_per_period = LLONG_MAX;
+    else
+      constraint.quota_per_period = strtoll(max_usage_str, NULL, 10);
+  }
+
+  /* Construct the path to the cpu.weight file */
+  snprintf(path, sizeof(path), "/sys/fs/cgroup/%.*s/cpu.weight", cgroup_size,
+           cgroup_trimmed);
+
+  /* Read cpu.weight */
+  if (uv__slurp(path, buf, sizeof(buf)) >= 0) {
+    sscanf(buf, "%u", &weight);
+    constraint.proportions = (double)weight / 100;
+  }
+
+  return constraint;
+}
+
+static char* uv__cgroup1_find_cpu_controller(const char* cgroup,
+                                             int* cgroup_size) {
+  char* cgroup_cpu;
+
+  /* Seek to the cpu controller line. */
+  cgroup_cpu = strstr(cgroup, ":cpu,");
+
+  if (cgroup_cpu != NULL) {
+    /* Determine the length of the mount path. */
+    cgroup_cpu = cgroup_cpu + sizeof(":cpu,") - 1;
+    *cgroup_size = (int)strcspn(cgroup_cpu, "\n");
+  }
+
+  return cgroup_cpu;
+}
+
+static uv__cpu_constraint uv__get_cgroupv1_constrained_cpu(const char* cgroup) {
+  uv__cpu_constraint constraint = {0, 0, 0};
+  char path[256];
+  char buf[BUF_SIZE];
+  unsigned int shares;
+  int cgroup_size;
+  char* cgroup_cpu;
+
+  cgroup_cpu = uv__cgroup1_find_cpu_controller(cgroup, &cgroup_size);
+
+  /* Construct the path to the cpu.max file */
+  snprintf(path, sizeof(path), "/sys/fs/cgroup/%.*s/cpu.cfs_quota_us",
+           cgroup_size, cgroup_cpu);
```

Comment:
Unsafe: cgroup_cpu can be NULL, cgroup_size can be uninitialized. You should handle the is-NULL case a few lines up.